### PR TITLE
[Xamarin.Android.Build.Tasks] warn exit code for "alias already exists"

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Context: 1ec865d0
Context: https://github.com/dotnet/msbuild/blob/74cbf25b3543a5c8c886a5a6659316f18dc49b8a/src/Utilities/ToolTask.cs#L370-L394

1ec865d0 attempted to fix an infrequent failure such as:

    Task "AndroidCreateDebugKey" (TaskId:407)
        Task Parameter:StorePass=android (TaskId:407)
        Task Parameter:ToolPath=C:\Users\AzDevOps\android-toolchain\jdk-11\bin (TaskId:407)
        Task Parameter:KeyAlias=androiddebugkey (TaskId:407)
        Task Parameter:Validity=10950 (TaskId:407)
        Task Parameter:KeyAlgorithm=RSA (TaskId:407)
        Task Parameter:StoreType=pkcs12 (TaskId:407)
        Task Parameter:KeyPass=android (TaskId:407)
        Task Parameter:Command=-genkeypair (TaskId:407)
        Task Parameter:KeyStore=C:\Users\AzDevOps\AppData\Local\Xamarin\Mono for Android\debug.keystore (TaskId:407)
        KeyTool : -genkeypair (TaskId:407)
                    C:\Users\AzDevOps\AppData\Local\Xamarin\Mono for Android\debug.keystore (TaskId:407)
        C:\Users\AzDevOps\android-toolchain\jdk-11\bin\keytool.exe -genkeypair -alias androiddebugkey -storepass android -keypass android -keystore "C:\Users\AzDevOps\AppData\Local\Xamarin\Mono for Android\debug.keystore" -dname "CN=Android Debug,O=Android,C=US" -keyalg RSA -validity 10950 -storetype pkcs12  (TaskId:407)
        Xamarin.Android.Common.targets(2169,2): error ANDKT0000: keytool error: java.lang.Exception: Key pair not generated, alias <androiddebugkey> already exists

The fix "worked" in downgrading the error:

    message ANDKT0000: keytool warning: java.lang.Exception: Key pair not generated, alias <androiddebugkey> already exists

But it did not prevent the overall exit code from triggering a failure:

    error MSB6006: "keytool" exited with code 1.

We can override `ToolTask.HandleTaskExecutionErrors()`, to return
`true` when the `alias <androiddebugkey> already exists` message is
encountered.

This should hopefully allow the build to succeed.

### CODE REVIEW

File: src/Xamarin.Android.Build.Tasks/Tasks/AndroidCreateDebugKey.cs [hunk 1]
```diff
@@ -27,16 +27,31 @@ protected override CommandLineBuilder CreateCommandLine ()
...
 			}
 
 			base.LogFromStandardError (text);
 		}
+
+		protected override bool HandleTaskExecutionErrors ()
+		{
+			// Warn for the exit code if we get:
+			// error: java.lang.Exception: Key pair not generated, alias <androiddebugkey> already exists
+			if (warnExitCode && ExitCode != 0) {
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tasks/AndroidCreateDebugKey.cs:48]: Why check for `ExitCode != 0` when the linked-to `HandleTaskExecutionErrors()` method:

1. has documentation which states:

    > This method will only be called with non-zero exitCode

 2. has a method body which *asserts* that `ExitCode != 0`

Comment (jonathanpeppers) [reply src/Xamarin.Android.Build.Tasks/Tasks/AndroidCreateDebugKey.cs]: I removed this check, thanks.

**Review by dellis1972 — APPROVED** [review]

File: src/Xamarin.Android.Build.Tasks/Tasks/AndroidCreateDebugKey.cs [hunk 2]
```diff
@@ -27,16 +27,31 @@ protected override CommandLineBuilder CreateCommandLine ()
...
 
+		bool warnExitCode = false;
+
 		protected override void LogFromStandardError (string text)
 		{
 			// Downgrade error message to warning:
 			// error: java.lang.Exception: Key pair not generated, alias <androiddebugkey> already exists
 			if (text.Contains ($"alias <{KeyAlias}> already exists")) {
 				text = text.Replace ("error:", "warning:");
+				warnExitCode = true;
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tasks/AndroidCreateDebugKey.cs:38]: Is it at all plausible for `keytool` to emit more than one error message?  A scenario that comes to mind is "what if" we get the "alias already exists" error *as well as* some other error.  With this logic, we'll ignore *all* the errors so long as one of them contains `$"alias <{KeyAlias}> already exists"`.

"Safer" might be:

```csharp
int errorMessageCount;

protected override void LogFromStandardError (string text)
{
	if (text.Contains ("error:", StringComparison.OrdinalIgnoreCase))
		errorMessageCount++;
	if text.Contains ($"alias <{KeyAlias}> already exists")) {
		text = text.Replace ("error:", "warning:");
		errorMessageCount--;
	}

	base.LogFromStandardError (text);
}

protected override bool HandleTaskExecutionErrors ()
{
	if (errorMessageCount > 0) {
		Log.LogWarningFromResources ("ToolTask.ToolCommandFailed", ToolExe, ExitCode);
		return true;
	}

	return base.HandleTaskExecutionErrors ();
}
```

Comment (jonathanpeppers) [reply src/Xamarin.Android.Build.Tasks/Tasks/AndroidCreateDebugKey.cs:38]: I added a check such as:
```csharp
if (warnExitCode && !Log.HasLoggedErrors) {
	Log.LogWarningFromResources ("ToolTask.ToolCommandFailed", ToolExe, ExitCode);
	return true;
}
```
So if anything logs an error before we get to the exit code, it won't change it to a warning.

I think that will do the same thing without incrementing an `int`.

File: src/Xamarin.Android.Build.Tasks/Tasks/AndroidCreateDebugKey.cs [hunk 3]
```diff
@@ -27,16 +27,31 @@ protected override CommandLineBuilder CreateCommandLine ()
...
 
 			base.LogFromStandardError (text);
 		}
+
+		protected override bool HandleTaskExecutionErrors ()
+		{
+			// Warn for the exit code if we get:
+			// error: java.lang.Exception: Key pair not generated, alias <androiddebugkey> already exists
+			if (warnExitCode && !Log.HasLoggedErrors) {
+				Log.LogWarningFromResources ("ToolTask.ToolCommandFailed", ToolExe, ExitCode);
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tasks/AndroidCreateDebugKey.cs:49]: Alas, this resource apparently doesn't exist:

https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4432778&view=ms.vss-test-web.build-test-results-tab&runId=18631396&resultId=100000&paneView=attachments

```
System.InvalidOperationException: The "AndroidCreateDebugKey" task has not registered its resources. In order to use the "TaskLoggingHelper.FormatResourceString()" method this task needs to register its resources either during construction, or via the "TaskResources" property.
  at Microsoft.Build.Shared.ErrorUtilities.ThrowInvalidOperation (System.String resourceName, System.Object[] args)
  at Microsoft.Build.Shared.ErrorUtilities.VerifyThrowInvalidOperation (System.Boolean condition, System.String resourceName, System.Object arg0)
  at Microsoft.Build.Utilities.TaskLoggingHelper.FormatResourceString (System.String resourceName, System.Object[] args)
  at Microsoft.Build.Utilities.TaskLoggingHelper.LogWarningFromResources (System.String subcategoryResourceName, System.String warningCode, System.String helpKeyword, System.String file, System.Int32 lineNumber, System.Int32 columnNumber, System.Int32 endLineNumber, System.Int32 endColumnNumber, System.String messageResourceName, System.Object[] messageArgs)
  at Microsoft.Build.Utilities.TaskLoggingHelper.LogWarningFromResources (System.String messageResourceName, System.Object[] messageArgs)
  at (wrapper remoting-invoke-with-check) Microsoft.Build.Utilities.TaskLoggingHelper.LogWarningFromResources(string,object[])
  at Xamarin.Android.Tasks.AndroidCreateDebugKey.HandleTaskExecutionErrors ()
  at Microsoft.Build.Utilities.ToolTask.Execute ()
```

Comment (jonathanpeppers) [reply src/Xamarin.Android.Build.Tasks/Tasks/AndroidCreateDebugKey.cs:55]: Trying locally this seems to fix it:
```
TaskResources = new ResourceManager ("Microsoft.Build.Utilities.Core.Strings", typeof (ToolTask).Assembly);
```

**Pull request merged**